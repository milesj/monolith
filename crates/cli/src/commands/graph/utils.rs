use super::dto::{GraphEdgeDto, GraphInfoDto, GraphNodeDto};
use crate::helpers::AnyError;
use moon_logger::info;
use moon_runner::DepGraph;
use moon_workspace::Workspace;
use petgraph::Graph;
use serde::Serialize;
use std::{collections::HashSet, env};
use tera::{Context, Tera};
use tiny_http::{Header, Response, Server};

const INDEX_HTML: &str = include_str!("graph.html.tera");
const LOG_TARGET: &str = "moon:graph::utils";

#[derive(Debug, Serialize)]
pub struct RenderContext {
    pub graph_data: String,

    pub js_url: String,
}

pub async fn setup_server() -> Result<(Server, Tera), AnyError> {
    let port = match env::var("PORT") {
        Ok(p) => p.parse::<u16>().unwrap(),
        Err(..) => 8000,
    };
    let host = match env::var("HOST") {
        Ok(h) => h,
        Err(..) => "127.0.0.1".to_string(),
    };
    let address = format!("{}:{}", host, port);
    let server = Server::http(address).unwrap();

    let tera = Tera::default();

    Ok((server, tera))
}

pub fn extract_nodes_and_edges_from_graph(graph: &Graph<String, ()>) -> GraphInfoDto {
    let edges = graph
        .raw_edges()
        .iter()
        .map(|e| GraphEdgeDto {
            source: e.source().index(),
            target: e.target().index(),
            id: format!("{}->{}", e.source().index(), e.target().index()),
        })
        .collect::<Vec<_>>();
    let mut nodes = HashSet::new();
    for edge in graph.raw_edges().iter() {
        let source = graph
            .node_weight(edge.source())
            .expect("Unable to get node")
            .clone();
        let target = graph
            .node_weight(edge.target())
            .expect("Unable to get node")
            .clone();
        nodes.insert(GraphNodeDto {
            id: edge.source().index(),
            label: source,
        });
        nodes.insert(GraphNodeDto {
            id: edge.target().index(),
            label: target,
        });
    }
    let nodes = nodes.into_iter().collect();
    GraphInfoDto { edges, nodes }
}

/// Get a serialized representation of the workspace graph.
pub async fn workspace_graph_repr(workspace: &Workspace) -> GraphInfoDto {
    let labeled_graph = workspace.projects.labeled_graph();
    extract_nodes_and_edges_from_graph(&labeled_graph)
}

/// Get a serialized representation of the dependency graph.
pub async fn dep_graph_repr(graph: &DepGraph) -> GraphInfoDto {
    let labeled_graph = graph.labeled_graph();
    extract_nodes_and_edges_from_graph(&labeled_graph)
}

pub fn respond_to_request(
    server: Server,
    tera: &mut Tera,
    graph: &GraphInfoDto,
) -> Result<(), AnyError> {
    info!(
        target: LOG_TARGET,
        r#"Starting server on "{}""#,
        server.server_addr()
    );
    for req in server.incoming_requests() {
        let response = match req.url() {
            "/graph-data" => {
                let mut response = Response::from_data(serde_json::to_string(graph)?);
                response.add_header(
                    Header::from_bytes(&b"Content-Type"[..], &b"application/json"[..]).unwrap(),
                );
                response
            }
            _ => {
                let graph_data = serde_json::to_string(graph)?;
                // Use the local version of the JS file when in development mode otherwise
                // the CDN url.
                let mut js_url = match cfg!(debug_assertions) {
                    // FIXME: We should create a separate module to store these constants
                    true => "http://localhost:5000".to_string(),
                    false => "https://cdn.com".to_string(),
                };
                js_url.push_str("/assets/index.js");
                let context = RenderContext { graph_data, js_url };
                let info = tera
                    .render_str(INDEX_HTML, &Context::from_serialize(&context)?)
                    .unwrap();
                let mut response = Response::from_data(info);
                response.add_header(
                    Header::from_bytes(&b"Content-Type"[..], &b"text/html"[..]).unwrap(),
                );
                response
            }
        };
        req.respond(response).unwrap_or_default();
    }
    Ok(())
}
