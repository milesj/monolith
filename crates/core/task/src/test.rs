use crate::{FileGroup, Target, Task};
use moon_config::TaskConfig;
use moon_utils::string_vec;
use rustc_hash::FxHashMap;

pub fn create_file_groups_config() -> FxHashMap<String, Vec<String>> {
    let mut map = FxHashMap::default();

    map.insert(
        String::from("static"),
        string_vec![
            "file.ts",
            "dir",
            "dir/other.tsx",
            "dir/subdir",
            "dir/subdir/another.ts",
        ],
    );

    map.insert(String::from("dirs_glob"), string_vec!["**/*"]);

    map.insert(String::from("files_glob"), string_vec!["**/*.{ts,tsx}"]);

    map.insert(String::from("globs"), string_vec!["**/*.{ts,tsx}", "*.js"]);

    map.insert(String::from("no_globs"), string_vec!["config.js"]);

    map
}

pub fn create_file_groups() -> FxHashMap<String, FileGroup> {
    let mut map = FxHashMap::default();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            string_vec![
                "file.ts",
                "dir",
                "dir/other.tsx",
                "dir/subdir",
                "dir/subdir/another.ts",
            ],
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", string_vec!["**/*"]),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", string_vec!["**/*.{ts,tsx}"]),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new("globs", string_vec!["**/*.{ts,tsx}", "*.js"]),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", string_vec!["config.js"]),
    );

    map
}

pub fn create_initial_task(config: Option<TaskConfig>) -> Task {
    Task::from_config(
        Target::new("project", "task").unwrap(),
        &config.unwrap_or_default(),
    )
    .unwrap()
}
