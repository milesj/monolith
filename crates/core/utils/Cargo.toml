[package]
name = "moon_utils"
version = "0.1.0"
edition = "2021"

[dependencies]
moon_constants = { path = "../constants" }
moon_error = { path = "../error" }
moon_logger = { path = "../logger" }
async-trait = { workspace = true }
cached = { workspace = true }
chrono = { version = "0.4.24", features = ["serde"] }
clean-path = "0.2.1"
dirs = "5.0.0"
dunce = "1.0.3"
ec4rs = "1.0.2"
humantime = "2.1.0"
lazy_static = { workspace = true }
md5 = "0.7.0"
pathdiff = "0.2.1"
regex = { workspace = true }
rustc-hash = { workspace = true }
semver = "1.0.17"
serde = { workspace = true }
shell-words = "1.1.0"
starbase_styles = { workspace = true }
tokio = { workspace = true }
wax = "0.5.0"

[dev-dependencies]
moon_test_utils = { path = "../test-utils" }
