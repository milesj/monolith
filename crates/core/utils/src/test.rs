use crate::path;
use crate::process::output_to_string;
use std::env;
use std::path::{Path, PathBuf};
use std::process::Command;

pub use assert_fs::TempDir;

pub fn run_git_command<P, F>(dir: P, handler: F)
where
    P: AsRef<Path>,
    F: FnOnce(&mut Command),
{
    let mut cmd = Command::new(if cfg!(windows) { "git.exe" } else { "git" });
    cmd.current_dir(dir.as_ref());

    handler(&mut cmd);

    let out = cmd.output().unwrap_or_else(|e| {
        panic!("{:#?}", e);
    });

    if !out.status.success() {
        println!("{}", output_to_string(&out.stdout));
        eprintln!("{}", output_to_string(&out.stderr));
    }
}

pub fn create_sandbox<T: AsRef<str>>(fixture: T) -> assert_fs::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = assert_fs::TempDir::new().unwrap();

    temp_dir
        .copy_from(get_fixtures_dir(fixture), &["**/*"])
        .unwrap();

    temp_dir
        .copy_from(get_fixtures_root(), &["shared-workspace.yml"])
        .unwrap();

    temp_dir
}

pub fn create_sandbox_with_git<T: AsRef<str>>(fixture: T) -> assert_fs::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = create_sandbox(fixture);

    if !temp_dir.path().join(".gitignore").exists() {
        temp_dir
            .child(".gitignore")
            .write_str("node_modules")
            .unwrap();
    }

    // Initialize a git repo so that VCS commands work
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["init", "--initial-branch", "master"]);
    });

    // We must also add the files to the index
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["add", "--all", "."]);
    });

    // And commit them... this seems like a lot of overhead?
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["commit", "-m", "Fixtures"])
            .env("GIT_AUTHOR_NAME", "moon tests")
            .env("GIT_AUTHOR_EMAIL", "fakeemail@moonrepo.dev")
            .env("GIT_COMMITTER_NAME", "moon tests")
            .env("GIT_COMMITTER_EMAIL", "fakeemail@moonrepo.dev");
    });

    temp_dir
}

pub fn get_fixtures_root() -> PathBuf {
    let mut root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    root.push("../../../tests/fixtures");

    path::normalize(root)
}

pub fn get_fixtures_dir<T: AsRef<str>>(dir: T) -> PathBuf {
    get_fixtures_root().join(dir.as_ref())
}

pub fn replace_fixtures_dir<T: AsRef<str>, P: AsRef<Path>>(value: T, dir: P) -> String {
    let dir_str = dir.as_ref().to_str().unwrap();

    // Replace both forward and backward slashes
    value
        .as_ref()
        .replace(dir_str, "<WORKSPACE>")
        .replace(&path::standardize_separators(dir_str), "<WORKSPACE>")
}

pub fn create_moon_command<T: AsRef<Path>>(path: T) -> assert_cmd::Command {
    let mut cmd = assert_cmd::Command::cargo_bin("moon").unwrap();
    cmd.current_dir(path);
    cmd.timeout(std::time::Duration::from_secs(90));
    // Let our code know were running tests
    cmd.env("MOON_TEST", "true");
    // Hide install output as it disrupts testing snapshots
    cmd.env("MOON_TEST_HIDE_INSTALL_OUTPUT", "true");
    // Standardize file system paths for testing snapshots
    cmd.env("MOON_TEST_STANDARDIZE_PATHS", "true");
    // Enable logging for code coverage
    cmd.env("MOON_LOG", "trace");
    cmd
}

pub fn get_assert_output(assert: &assert_cmd::assert::Assert) -> String {
    get_assert_stdout_output(assert) + &get_assert_stderr_output_clean(assert)
}

pub fn get_assert_stderr_output_clean(assert: &assert_cmd::assert::Assert) -> String {
    let mut output = String::new();

    // We need to always show logs for proper code coverage,
    // but this breaks snapshots, and as such, we need to manually
    // filter out log lines and env vars!
    for line in get_assert_stderr_output(assert).split('\n') {
        if !line.starts_with("[error")
            && !line.starts_with("[ warn")
            && !line.starts_with("[ info")
            && !line.starts_with("[debug")
            && !line.starts_with("[trace")
            && !line.starts_with("  MOON_")
            && !line.starts_with("  NODE_")
        {
            output.push_str(line);
            output.push('\n');
        }
    }

    output
}

pub fn get_assert_stderr_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stderr.to_owned()).unwrap()
}

pub fn get_assert_stdout_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stdout.to_owned()).unwrap()
}

pub fn debug_sandbox_files(dir: &Path) {
    for entry in std::fs::read_dir(dir).unwrap() {
        let path = entry.unwrap().path();

        if path.is_dir() {
            debug_sandbox_files(&path);
        } else {
            println!("- {}", path.to_string_lossy());
        }
    }
}

pub fn debug_sandbox(fixture: &assert_fs::TempDir, assert: &assert_cmd::assert::Assert) {
    // List all files in the sandbox
    println!("sandbox:");
    debug_sandbox_files(fixture.path());
    println!("\n");

    // Debug outputs
    println!("stdout:\n{}\n", get_assert_stdout_output(assert));
    println!("stderr:\n{}\n", get_assert_stderr_output(assert));
    println!("status: {:#?}", assert.get_output().status);
}
