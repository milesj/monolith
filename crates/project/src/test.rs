use crate::errors::ProjectError;
use crate::file_group::FileGroup;
use crate::target::Target;
use crate::task::Task;
use crate::token::{TokenResolver, TokenSharedData};
use moon_config::TaskConfig;
use std::collections::HashMap;
use std::path::Path;

pub fn create_file_groups_config() -> HashMap<String, Vec<String>> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        vec![
            "file.ts".to_owned(),
            "dir".to_owned(),
            "dir/other.tsx".to_owned(),
            "dir/subdir".to_owned(),
            "dir/subdir/another.ts".to_owned(),
        ],
    );

    map.insert(String::from("dirs_glob"), vec!["**/*".to_owned()]);

    map.insert(String::from("files_glob"), vec!["**/*.{ts,tsx}".to_owned()]);

    map.insert(
        String::from("globs"),
        vec!["**/*.{ts,tsx}".to_owned(), "*.js".to_owned()],
    );

    map.insert(String::from("no_globs"), vec!["config.js".to_owned()]);

    map
}

pub fn create_file_groups(project_root: &Path) -> HashMap<String, FileGroup> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            vec![
                "file.ts".to_owned(),
                "dir".to_owned(),
                "dir/other.tsx".to_owned(),
                "dir/subdir".to_owned(),
                "dir/subdir/another.ts".to_owned(),
            ],
            project_root,
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", vec!["**/*".to_owned()], project_root),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", vec!["**/*.{ts,tsx}".to_owned()], project_root),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new(
            "globs",
            vec!["**/*.{ts,tsx}".to_owned(), "*.js".to_owned()],
            project_root,
        ),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", vec!["config.js".to_owned()], project_root),
    );

    map
}

pub fn create_expanded_task(
    workspace_root: &Path,
    project_root: &Path,
    config: Option<TaskConfig>,
) -> Result<Task, ProjectError> {
    let mut task = Task::from_config(
        Target::format("project", "task").unwrap(),
        &config.unwrap_or_default(),
    );
    let file_groups = create_file_groups(project_root);
    let metadata = TokenSharedData::new(&file_groups, workspace_root, project_root);

    task.expand_inputs(TokenResolver::for_inputs(&metadata))?;
    task.expand_outputs(TokenResolver::for_outputs(&metadata))?;
    task.expand_args(TokenResolver::for_args(&metadata))?; // Must be last

    Ok(task)
}
