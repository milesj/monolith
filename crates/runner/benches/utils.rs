use moon_contract::Platformable;
use moon_platform_node::NodePlatform;
use moon_platform_system::SystemPlatform;
use moon_project_graph::ProjectGraph;
use moon_runner::DepGraph;
use moon_task::Target;
use moon_workspace::Workspace;

pub fn setup_platforms(workspace: &mut Workspace) {
    workspace
        .projects
        .register_platform(Box::new(SystemPlatform::default()))
        .unwrap();

    workspace
        .projects
        .register_platform(Box::new(NodePlatform::default()))
        .unwrap();
}

pub fn setup_dep_graph(project_graph: &ProjectGraph) -> DepGraph {
    let mut dep_graph = DepGraph::default();

    dep_graph
        .run_target(Target::parse("base:base").unwrap(), project_graph, &None)
        .unwrap();

    dep_graph
        .run_target(
            Target::parse("depsA:dependencyOrder").unwrap(),
            project_graph,
            &None,
        )
        .unwrap();

    dep_graph
        .run_target(
            Target::parse("node:standard").unwrap(),
            project_graph,
            &None,
        )
        .unwrap();

    dep_graph
        .run_target(Target::parse("system:bash").unwrap(), project_graph, &None)
        .unwrap();

    dep_graph
        .run_target(
            Target::parse("targetScopeB:self").unwrap(),
            project_graph,
            &None,
        )
        .unwrap();

    dep_graph
}
