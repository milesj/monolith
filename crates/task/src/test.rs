use crate::{FileGroup, Target, Task, TaskError, TokenResolver, TokenSharedData};
use moon_config::{ProjectConfig, TaskConfig};
use moon_utils::string_vec;
use std::collections::HashMap;
use std::path::Path;

pub fn create_file_groups_config() -> HashMap<String, Vec<String>> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        string_vec![
            "file.ts",
            "dir",
            "dir/other.tsx",
            "dir/subdir",
            "dir/subdir/another.ts",
        ],
    );

    map.insert(String::from("dirs_glob"), string_vec!["**/*"]);

    map.insert(String::from("files_glob"), string_vec!["**/*.{ts,tsx}"]);

    map.insert(String::from("globs"), string_vec!["**/*.{ts,tsx}", "*.js"]);

    map.insert(String::from("no_globs"), string_vec!["config.js"]);

    map
}

pub fn create_file_groups() -> HashMap<String, FileGroup> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            string_vec![
                "file.ts",
                "dir",
                "dir/other.tsx",
                "dir/subdir",
                "dir/subdir/another.ts",
            ],
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", string_vec!["**/*"]),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", string_vec!["**/*.{ts,tsx}"]),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new("globs", string_vec!["**/*.{ts,tsx}", "*.js"]),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", string_vec!["config.js"]),
    );

    map
}

pub fn create_initial_task(config: Option<TaskConfig>) -> Task {
    Task::from_config(
        Target::format("project", "task").unwrap(),
        &config.unwrap_or_default(),
    )
}

pub fn create_expanded_task(
    workspace_root: &Path,
    project_root: &Path,
    config: Option<TaskConfig>,
) -> Result<Task, TaskError> {
    let mut task = create_initial_task(config);
    let file_groups = create_file_groups();
    let project_config = ProjectConfig::new(project_root);
    let metadata =
        TokenSharedData::new(&file_groups, workspace_root, project_root, &project_config);

    task.expand_env(&metadata)?;
    task.expand_inputs(TokenResolver::for_inputs(&metadata))?;
    task.expand_outputs(TokenResolver::for_outputs(&metadata))?;
    task.expand_args(TokenResolver::for_args(&metadata))?; // Must be last

    Ok(task)
}
