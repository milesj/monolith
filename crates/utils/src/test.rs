use crate::path;
use std::env;
use std::path::{Path, PathBuf};

pub fn create_fixtures_sandbox(dir: &str) -> assert_fs::fixture::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = assert_fs::fixture::TempDir::new().unwrap();

    temp_dir
        .copy_from(get_fixtures_dir(dir), &["**/*"])
        .unwrap();

    temp_dir
}

pub fn get_fixtures_dir(dir: &str) -> PathBuf {
    get_fixtures_root().join(dir)
}

pub fn get_fixtures_root() -> PathBuf {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("../../tests/fixtures");

    path.canonicalize().unwrap()
}

pub fn replace_fixtures_dir(value: &str, dir: &Path) -> String {
    value.replace(dir.to_str().unwrap(), "<WORKSPACE>")
}

// We need to do this so slashes are accurate and always forward
pub fn wrap_glob(path: &Path) -> PathBuf {
    PathBuf::from(path::normalize_glob(path))
}

pub fn create_moon_command(fixture: &str) -> assert_cmd::Command {
    let mut cmd = create_moon_command_in(&get_fixtures_dir(fixture));
    // Never cache in these tests since they're not in a sandbox
    cmd.env("MOON_CACHE", "off");
    cmd
}

pub fn create_moon_command_in(path: &Path) -> assert_cmd::Command {
    let mut cmd = assert_cmd::Command::cargo_bin("moon").unwrap();
    cmd.current_dir(path);
    // Let our code know were running tests
    cmd.env("MOON_TEST", "true");
    // Hide install output as it disrupts testing snapshots
    cmd.env("MOON_TEST_HIDE_INSTALL_OUTPUT", "true");
    // Standardize file system paths for testing snapshots
    cmd.env("MOON_TEST_STANDARDIZE_PATHS", "true");
    // Uncomment for debugging
    // cmd.arg("--logLevel");
    // cmd.arg("trace");
    cmd
}

pub fn get_assert_output(assert: &assert_cmd::assert::Assert) -> String {
    get_assert_stdout_output(assert) + &get_assert_stderr_output(assert)
}

pub fn get_assert_stderr_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stderr.to_owned()).unwrap()
}

pub fn get_assert_stdout_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stdout.to_owned()).unwrap()
}
