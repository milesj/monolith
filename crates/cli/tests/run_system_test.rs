mod utils;

use insta::assert_snapshot;
use moon_utils::test::{create_moon_command, create_sandbox_with_git, get_assert_output};
use utils::get_path_safe_output;

#[cfg(not(windows))]
mod unix {
    use super::*;

    #[test]
    fn handles_echo() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:echo")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn handles_ls() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:ls")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn runs_bash_script() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:bash")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn handles_process_exit_zero() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:exitZero")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn handles_process_exit_nonzero() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:exitNonZero")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn passes_args_through() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:passthroughArgs")
            .arg("--")
            .arg("-aBc")
            .arg("--opt")
            .arg("value")
            .arg("--optCamel=value")
            .arg("foo")
            .arg("'bar baz'")
            .arg("--opt-kebab")
            .arg("123")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn sets_env_vars() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:envVars")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn inherits_moon_env_vars() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:envVarsMoon")
            .assert();

        assert_snapshot!(get_path_safe_output(&assert, fixture.path()));
    }

    #[test]
    fn runs_from_project_root() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:runFromProject")
            .assert();

        assert_snapshot!(get_path_safe_output(&assert, fixture.path()));
    }

    #[test]
    fn runs_from_workspace_root() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:runFromWorkspace")
            .assert();

        assert_snapshot!(get_path_safe_output(&assert, fixture.path()));
    }

    #[test]
    fn retries_on_failure_till_count() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("unix:retryCount")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    mod passthrough {
        use super::*;

        #[test]
        fn doesnt_pass_to_non_primary() {
            let fixture = create_sandbox_with_git("system");

            let assert = create_moon_command(fixture.path())
                .arg("run")
                .arg("passthroughArgs:b")
                .arg("--")
                .arg("-aBc")
                .assert();

            assert_snapshot!(get_assert_output(&assert));
        }
    }
}

#[cfg(windows)]
mod system_windows {
    use super::*;

    #[test]
    fn runs_bat_script() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:bat")
            .assert();

        assert_snapshot!(get_path_safe_output(&assert, fixture.path()));
    }

    #[test]
    fn handles_process_exit_zero() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:exitZero")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn handles_process_exit_nonzero() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:exitNonZero")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn passes_args_through() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:passthroughArgs")
            .arg("--")
            .arg("-aBc")
            .arg("--opt")
            .arg("value")
            .arg("--optCamel=value")
            .arg("foo")
            .arg("'bar baz'")
            .arg("--opt-kebab")
            .arg("123")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn sets_env_vars() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:envVars")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }

    #[test]
    fn inherits_moon_env_vars() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:envVarsMoon")
            .assert();

        assert_snapshot!(get_path_safe_output(&assert, fixture.path()));
    }

    #[test]
    fn runs_from_project_root() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:runFromProject")
            .assert();

        assert_snapshot!(get_path_safe_output(&assert, fixture.path()));
    }

    #[test]
    fn runs_from_workspace_root() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:runFromWorkspace")
            .assert();

        assert_snapshot!(get_path_safe_output(&assert, fixture.path()));
    }

    #[test]
    fn retries_on_failure_till_count() {
        let fixture = create_sandbox_with_git("system");

        let assert = create_moon_command(fixture.path())
            .arg("run")
            .arg("windows:retryCount")
            .assert();

        assert_snapshot!(get_assert_output(&assert));
    }
}
