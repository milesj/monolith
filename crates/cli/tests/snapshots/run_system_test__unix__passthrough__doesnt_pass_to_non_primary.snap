---
source: crates/cli/tests/run_system_test.rs
assertion_line: 166
expression: get_assert_output(&assert)
---

 ERROR  No project has been configured with the ID passthroughArgs.



